WcaOnRails/lib/finish_unfinished_persons.rb

```diff
@@ -0,0 +1,211 @@
+                 .group(:personId, :personName, :competitionId, :countryId)
+                 .order(:personName)
+  end
+
+  def self.search_persons(competition_ids = nil)
+    unfinished_person_results = self.unfinished_results_scope(competition_ids)
+
+    unfinished_persons = []
+    available_id_spots = {} # to make sure that all of the newcomer IDs that we're creating in one batch are unique among each other
+
+    @persons_cache = nil
+
+    unfinished_person_results.each do |res|
+      next if unfinished_persons.length >= MAX_PER_BATCH
+
+      competition_year = res.competition.year
+      inbox_dob = res.inbox_person&.dob
+
+      roman_name = self.extract_roman_name res.person_name
+      sanitized_roman_name = self.remove_accents roman_name
+      name_parts = sanitized_roman_name.gsub(/[^a-zA-Z ]/, '').upcase.split
+
+      last_name = name_parts[-1]
+      rest_of_name = name_parts[...-1].join
+
+      padded_rest_of_name = rest_of_name.ljust WCA_QUARTER_ID_LENGTH, WCA_ID_PADDING
+      letters_to_shift = [0, WCA_QUARTER_ID_LENGTH - last_name.length].max
+
+      semi_id = nil
+      cleared_id = false
+
+      until cleared_id || letters_to_shift > WCA_QUARTER_ID_LENGTH
+        quarter_id = last_name[...(WCA_QUARTER_ID_LENGTH - letters_to_shift)] + padded_rest_of_name[...letters_to_shift]
+        semi_id = competition_year.to_s + quarter_id
+
+        unless available_id_spots.key?(semi_id)
+          # TODO: Merge with anonymization code
+          last_id_taken = Person.where('wca_id LIKE ?', "#{semi_id}__")
+                                .order(wca_id: :desc)
+                                .pluck(:wca_id)
+                                .first
+
+          if last_id_taken.present?
+            counter = last_id_taken[(4 + WCA_QUARTER_ID_LENGTH)..].to_i
+          else
+            counter = 0
+          end
+
+          available_id_spots[semi_id] = 99 - counter
+        end
+
+        if available_id_spots.key?(semi_id)
+          available_id_spots[semi_id] -= 1
+          cleared_id = true
+        else
+          letters_to_shift += 1
+        end
+      end
+
+      unless cleared_id
+        raise "Could not compute a semi-id for #{res.person_name}"
+      end
+
+      similar_persons = compute_similar_persons(res)
+
+      unfinished_persons.push({
+                                person_id: res.person_id,
+                                person_name: res.person_name,
+                                country_id: res.country_id,
+                                competition_id: res.competition_id,
+                                person_dob: inbox_dob,
+                                computed_semi_id: semi_id,
+                                similar_persons: similar_persons,
+                              })
+    end
+
+    unfinished_persons
+  end
+
+  def self.extract_roman_name(person_name)
+    name_matches = person_name.match(/(.*)\((.*)\)$/)
```

Comment:
Would be nice to have a comment here that we store People with non Roman name in the form `Roman Name (non roman name)` because the regex looks confusing otherwise